keps/sig-network/20190324-remove-kube-proxy-autocleanup.md

```diff
@@ -0,0 +1,74 @@
+---
+kep-number: 0
+title: Remove kube-proxy's automatic clean up logic
+authors:
+  - "@bowei"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - "@andrewsykim"
+approvers:
+  - "@bowei"
+  - "@thockin"
+editor:
+creation-date: 2018-03-24
+last-updated: 2018-03-24
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Remove kube-proxy's automatic clean up logic
+
+## Summary
+
+Remove the automatic network rule cleanup from kube-proxy.
+Only clean up rules on boot if the `--cleanup` flag is present.
+
+## Motivation
+
+kube-proxy's rule cleanup has a substantial performance impact,
+and is not normally necessary.
+
+### Goals
+
+* Reduce the typical kube-proxy startup time,
+by avoiding unnecessary cleanup.
+
+## Proposal
+
+Add the flag `--cleanup` to kube-proxy,
```

Comment:
We actually have that flag already.

What we want to do is treat that flag as the ONLY reason to consider cleaning up.

Now one could ask, should we even bother trying to clean up AT ALL?  I'm not sure of any use case for it.  Maybe @bowei or @m1093782566 or @dcbw  or @danwinship  or others do?

What if we stop cleaning up at startup NOW (i.e. fix the bug), deprecate those flags NOW, wait 6 months, and then get rid of that logic entirely?

Reply:
I'm fine with saying we don't support switching proxy modes without rebooting.

If we support upgrading kubernetes without rebooting then the proxy may need to clean up its own old rules in some cases (eg, https://github.com/kubernetes/kubernetes/commit/780d595). That's totally different code from what the PR this KEP is based on was modifying, but this KEP is currently phrased in a way that makes it sound like it wants to get rid of *all* cleanup, not just Other People's Proxy cleanup. (If we don't support upgrades without rebooting then we don't need to worry about this either though.)